demos/GMW_demo/gmw_OT.js

```diff
@@ -0,0 +1,179 @@
+/**
+OT
+**/
+$ = require('jquery-deferred');
+// party i :  receive aibj+ajbi from party j;
+// get choose then ^aibi
+// then share result to party i, j sxor to open
+
+
+
+function gmw_and(jiff,share1,share2) {
+  if (!(share1.jiff === share2.jiff)) {
+    throw new Error('shares do not belong to the same instance (^)');
+  }
+  if (!share1.jiff.helpers.Zp_equals(share1, share2)) {
+    throw new Error('shares must belong to the same field (^)');
+  }
+  if (!share1.jiff.helpers.array_equals(share1.holders, share2.holders)) {
+    throw new Error('shares must be held by the same parties (^)');
+  }
+
+
+
+  // XOR the two shares when ready locally
+  var ready = function () {
+    var final_deferred = new jiff.helpers.Deferred();
+    var final_promise = final_deferred.promise;
+    var wi= share1.value & share2.value;
+    var csecret=[];
+    csecret.push(share1.value);
+    csecret.push(share2.value);
+    // receivinglist of OT msg
+    var recls=[];
+    var i;
+    for (i=1;i<=jiff.party_count;i++) {
+      if (i<jiff.id) {
+        recls.push(i);
+      }
+    }
+
+    var share_id ;
+    if (share_id==null) {
+      share_id= jiff.counters.gen_op_id2('otshare', share1.holders, share1.holders);
+      //console.log(jiff.id,'first op_gen1',share_id);
+    }
+    var my_count;// deferlist
+    if (typeof (my_count) === 'undefined') {
+      my_count= initial_count(jiff,share_id);
+    }
+    //console.log(jiff.id,'no share',share_id,my_count);
+    // generate OT msg
+    var four_opts={};
+    four_opts['opts']=OTGate(csecret);// jason object
+    four_opts['sender_id']=jiff.id;
+    four_opts['op_id']=share_id;
+    four_opts = jiff.hooks.execute_array_hooks('beforeOperation', [jiff, 'open', four_opts], 2);
+    var mymsg=JSON.stringify(four_opts);
+    //console.log(jiff.id,share_id,'send OT msg to  ',recls);
+    var help1=[];
+    help1.push(wi);
+    jiff.deferreds[share_id][jiff.id].resolve(help1);//wi resolve local wi; when jiff.id===jiff.party_count, simple return
+    //console.log(jiff.id,share_id,'wi=',wi);
+    // OT send process
+    // if i>j
+    // send OT msg to party_id list,eg.recls
+    jiff.emit('OT',recls,mymsg,true);
+
+    // OT receive process
+    for ( var ssid=jiff.id+1;ssid<=jiff.party_count;ssid++) {
+
+      jiff.listen('OT',function (ssid,msg) {
+        msg=JSON.parse(msg);
+        var sendId=msg['sender_id'];
+        var opId=msg['op_id'];
+        var opts=msg['opts'];
+        //console.log(jiff.id,'get ot msg from',sendId,msg['op_id'],'!!',share_id,msg,opts);
+        if (jiff.deferreds[opId] == null) {
+          jiff.deferreds[opId] = {};
+        }
+        if (jiff.deferreds[opId][sendId] == null) { // not ready, setup a deferred
+          jiff.deferreds[opId][sendId] = new jiff.helpers.Deferred();
+        }
+        jiff.deferreds[opId][sendId].resolve(opts);
+      });
+    }
+
+    Promise.all(my_count).then(function (v) {
```

Comment:
the array of promises that is awaited here can be constructed from a similar loop to the one above, except just by pushing the result of OT.receive() into an array. (attaching the pseudocode file we wrote yesterday, that structure should be almost identical to what you do here, except with appropriate parameters passed to OT.receive/send 
[pseudocode.txt](https://github.com/multiparty/jiff/files/4860042/pseudocode.txt)